contracts/Staking.sol

```diff
@@ -64,36 +70,62 @@ contract Staking {
         if (checkpoints.length == 0) {
-            checkpoints.push(Checkpoint(currentEpoch, 0));
-            checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
+            checkpoints.push(Checkpoint(currentEpoch, currentMultiplier, 0, amount));
+
+            // next epoch => multiplier is 1, epoch deposits is 0
+            checkpoints.push(Checkpoint(currentEpoch + 1, BASE_MULTIPLIER, amount, 0));
         } else {
             uint256 last = checkpoints.length - 1;
 
             // the last action happened in an older epoch (e.g. a deposit in epoch 3, current epoch is >=5)
             if (checkpoints[last].epochId < currentEpoch) {
-                checkpoints.push(Checkpoint(currentEpoch, checkpoints[last].balance));
-                checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
+                checkpoints.push(Checkpoint(currentEpoch, currentMultiplier, getCheckpointEffectiveBalance(checkpoints[last]), amount));
+                checkpoints.push(Checkpoint(currentEpoch + 1, BASE_MULTIPLIER, balances[msg.sender][tokenAddress], 0));
             }
             // the last action happened in the previous epoch
             else if (checkpoints[last].epochId == currentEpoch) {
-                checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
+                checkpoints[last].multiplier = computeNewMultiplier(
+                    getCheckpointBalance(checkpoints[last]),
+                    checkpoints[last].multiplier,
+                    amount,
+                    currentMultiplier
+                );
+                checkpoints[last].newDeposits = checkpoints[last].newDeposits.add(amount);
+
+                checkpoints.push(Checkpoint(currentEpoch + 1, BASE_MULTIPLIER, balances[msg.sender][tokenAddress], 0));
             }
             // the last action happened in the current epoch
             else {
-                checkpoints[checkpoints.length - 1].balance = balances[msg.sender][tokenAddress];
+                if (last >= 1 && checkpoints[last - 1].epochId == currentEpoch) {
+                    checkpoints[last - 1].multiplier = computeNewMultiplier(
+                        getCheckpointBalance(checkpoints[last - 1]),
+                        checkpoints[last - 1].multiplier,
+                        amount,
+                        currentMultiplier
+                    );
+                    checkpoints[last - 1].newDeposits = checkpoints[last - 1].newDeposits.add(amount);
+                }
+
+                checkpoints[last].startBalance = balances[msg.sender][tokenAddress];
             }
         }
+
+        uint256 balanceAfter = getEpochUserBalance(msg.sender, tokenAddress, currentEpoch);
+
+        poolSize[tokenAddress][currentEpoch].size = poolSize[tokenAddress][currentEpoch].size.add(balanceAfter.sub(balanceBefore));
```

Comment:
`balanceBefore` and `balanceAfter` should be fetched with `getCheckpointEffectiveBalance`

Reply:
all good, `getEpochUserBalance` returns effective balance.